proc/put/put.go

```diff
@@ -81,116 +84,284 @@ func (p *Proc) eval(in *zng.Record) ([]zng.Value, error) {
 	return vals, nil
 }
 
-func (p *Proc) buildRule(inType *zng.TypeRecord, vals []zng.Value) (*putRule, error) {
-	n := len(inType.Columns)
-	cols := make([]zng.Column, n, n+len(p.clauses))
-	copy(cols, inType.Columns)
-	clauseTypes := make([]clauseType, len(p.clauses))
-	var nreplace int
-	replace := make([]int, n)
-	for k := range replace {
-		replace[k] = -1
+// A op is a recursive data structure encoding a series of steps to be
+// carried out to construct an output record from an input record and
+// a slice of evaluated clauses.
+type op struct {
+	op        step
+	index     int
+	container bool
+	record    []op // for op == record
+}
+
+func (s *op) append(step op) {
+	s.record = append(s.record, step)
+}
+
+type step int
+
+const (
+	fromInput  step = iota // copy field from input record
+	fromClause             // copy field from put assignment
+	root                   // values are being copied to the root record (put .=)
+	record                 // recurse into record below us
+)
+
+func (s op) build(in zcode.Bytes, b *zcode.Builder, vals []zng.Value) error {
+	if s.op == root {
+		iter := zcode.Iter(vals[s.index].Bytes)
+		for !iter.Done() {
+			bytes, c, err := iter.Next()
+			if err != nil {
+				return err
+			}
+			if c {
+				b.AppendContainer(bytes)
+			} else {
+				b.AppendPrimitive(bytes)
+			}
+		}
```

Comment:
Instead of copying `vals[s.index].Byte` using `b` like this, maybe change the method signature so you can use `copy`?
```go
func (s op) build(in zcode.Bytes, b *zcode.Builder, vals []zng.Value) error {
	switch s.op {
	case root:
		bytes := make(zcode.Bytes, len(vals[s.index].Bytes))
		copy(bytes, vals[s.index].Bytes)
		return bytes, nil
	case record:
		b.Reset()
		if err := s.buildRecord(in, b, vals); err != nil {
			return nil, err
		}
		return b.Bytes(), nil
	default:
		// top-level op should be root or record
		panic(fmt.Sprintf("put: unexpected op %v", op))
}
```

Reply:
👍 (and switch is also more readable).